discovery/syncer.go

```diff
@@ -929,69 +946,80 @@ func (g *GossipSyncer) replyChanRangeQuery(query *lnwire.QueryChannelRange) erro
 	// TODO(roasbeef): means can't send max uint above?
 	//  * or make internal 64
 
-	// In the base case (no actual response) the first block and last block
-	// will match those of the query. In the loop below, we'll update these
-	// two variables incrementally with each chunk to properly compute the
-	// starting block for each response and the number of blocks in a
-	// response.
-	firstBlockHeight := startBlock
-	lastBlockHeight := endBlock
-
-	numChannels := int32(len(channelRange))
-	numChansSent := int32(0)
+	// We'll send our response in a streaming manner, chunk-by-chunk. We do
+	// this as there's a transport message size limit which we'll need to
+	// adhere to.
+	var channelChunk, prevChannelChunk []lnwire.ShortChannelID
 	for {
-		// We'll send our this response in a streaming manner,
-		// chunk-by-chunk. We do this as there's a transport message
-		// size limit which we'll need to adhere to.
-		var channelChunk []lnwire.ShortChannelID
-
-		// We know this is the final chunk, if the difference between
-		// the total number of channels, and the number of channels
-		// we've sent is less-than-or-equal to the chunk size.
-		isFinalChunk := (numChannels - numChansSent) <= g.cfg.chunkSize
-
-		// If this is indeed the last chunk, then we'll send the
-		// remainder of the channels.
-		if isFinalChunk {
-			channelChunk = channelRange[numChansSent:]
+		// In the base case (no actual response), the first block and
+		// last block will match those of the query.
+		firstBlockHeight := startBlock
+		lastBlockHeight := endBlock
+
+		nextIdx := 0
+		for i, channels := range channelRange {
+			// If our current chunk already contains channels for
+			// some blocks, and adding the next block to the chunk
+			// results in exceeding the chunk size, then we'll send
+			// the current chunk as is and continue with a new
+			// chunk.
+			if len(channelChunk) > 0 &&
+				int32(len(channelChunk)+len(channels)) > g.cfg.chunkSize {
+				nextIdx = i
+				break
+			}
+
+			// Otherwise, limit the chunk size, which may result in
+			// channels being left out of the reply. This is fine as
+			// all we need to ensure for a proper reply is that we
+			// include as many channels as possible for one block
+			// within one reply.
+			chunkSize := int32(len(channels))
+			if chunkSize > g.cfg.chunkSize {
+				chunkSize = g.cfg.chunkSize
+			}
+			channelChunk = append(channelChunk, channels[:chunkSize]...)
+		}
+
+		// If we have a chunk to send, we'll remove the channels
+		// included in this chunk from our range and modify our reply
+		// height bounds.
+		if len(channelChunk) > 0 {
+			if nextIdx == 0 {
+				// Edge case to handle our last set of channels
+				// being included in the last chunk.
+				channelRange = nil
+			} else {
+				channelRange = channelRange[nextIdx:]
```

Comment:
Ah so it's re-sliced instead of indexed explicitly from the start...I think the mutation here can make the new logic a bit harder to follow. Also given the index `i` is used over this re-sliced backing slice, the precise implications aren't obvious at a glance.